felix/idalloc/index_allocator.go

```diff
@@ -26,24 +27,72 @@ type IndexRange struct {
 	Min, Max int
 }
 
+func (r IndexRange) contains(a int) bool {
+	return r.Min <= a && r.Max >= a
+}
+
+// ByMaxIndex sorts collections of IndexRange structs in order of their starting/lower index
+type ByMaxIndex []IndexRange
+
+// Len is the number of indexranges in the collection
+func (i ByMaxIndex) Len() int { return len(i) }
+
+// Less reports whether the element with index a
+// must sort before the element with index b.
+func (i ByMaxIndex) Less(a, b int) bool { return i[a].Max < i[b].Max }
+
+// Swap swaps the elements with indexes a and b.
+func (i ByMaxIndex) Swap(a, b int) { i[a], i[b] = i[b], i[a] }
+
 type IndexAllocator struct {
 	indexStack *stack.Stack
+	exclusions []IndexRange
 }
 
-func NewIndexAllocator(indexRange IndexRange) *IndexAllocator {
+// NewIndexAllocator returns an index allocator from the provided indexRanges
+// any indices falling within the specified exclusions will not be returned, even if designated by indexRanges
+func NewIndexAllocator(indexRanges []IndexRange, exclusions []IndexRange) *IndexAllocator {
+	// sort index ranges in descending order of their Max bound
+	if len(indexRanges) > 1 {
+		sort.Sort(sort.Reverse(ByMaxIndex(indexRanges)))
+	}
+
 	r := &IndexAllocator{
 		indexStack: stack.New(),
+		exclusions: exclusions,
 	}
-	// Push in reverse order so that the lowest index will come out first.
-	for i := indexRange.Max; i >= indexRange.Min; i-- {
-		r.indexStack.Push(i)
+
+	var lowestIndex int
+	for j, indexRange := range indexRanges {
+		if j == 0 {
+			// keep track of the most recent Min to prevent an overlapping range
+			// from creating duplicate indices in the final index stack
+			lowestIndex = indexRange.Max + 1
+		}
+
+		// Push in reverse order so that the lowest index will come out first.
+		populating:
+		for i := indexRange.Max; i >= indexRange.Min; i-- {
+			// skip overlapping range indices
+			if i >= lowestIndex {
+				continue
+			}
+			// skip exclusions
+			for _, excl := range exclusions {
+				if excl.contains(i) {
+					continue populating
```

Comment:
now ignoring routetables here, based on `exclusions` passed to the indexAllocator - I also log.Info when the same is noticed in validator and param_types.go 
CC @fasaxc